File: LeetcodeProblems/blind-75/dp/house-robber-ii/solution.java
// House Robber II - LeetCode #213

class Solution {
    public int rob(int[] nums) {
        if (nums.length == 0) return 0;
        if (nums.length == 1) return nums[0];
        
        // Rob without first house
        int max1 = robLinear(nums, 1, nums.length - 1);
        // Rob without last house
        int max2 = robLinear(nums, 0, nums.length - 2);
        
        return Math.max(max1, max2);
    }
    
    private int robLinear(int[] nums, int start, int end) {
        if (start > end) return 0;
        
        int prev = 0;
        int curr = 0;
        
        for (int i = start; i <= end; i++) {
            int temp = Math.max(prev + nums[i], curr);
            prev = curr;
            curr = temp;
        }
        
        return curr;
    }
}

File: LeetcodeProblems/blind-75/dp/house-robber-ii/solution.js
// House Robber II - LeetCode #213

/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function(nums) {
    if (nums.length === 0) return 0;
    if (nums.length === 1) return nums[0];
    
    // Rob without first
    const max1 = robLinear(nums, 1, nums.length - 1);
    // Rob without last
    const max2 = robLinear(nums, 0, nums.length - 2);
    
    return Math.max(max1, max2);
};

function robLinear(nums, start, end) {
    if (start > end) return 0;
    
    let prev = 0;
    let curr = 0;
    
    for (let i = start; i <= end; i++) {
        const temp = Math.max(prev + nums[i], curr);
        prev = curr;
        curr = temp;
    }
    
    return curr;
}
